README
This is acer_acpi version 0.11.1

### THIS IS VERY MUCH A WORK IN PROGRESS.  NO WARRANTY GIVEN.  IF YOU
### BREAK YOUR LAPTOP USING THIS DRIVER, YOU GET TO KEEP THE PIECES AND
### I WILL ACCEPT ABSOLUTELY NO RESPONSIBILITY.

acer_acpi is a driver to allow you to control various aspects of your Acer
laptop hardware using the proc filesystem.  It was originally developed by Mark
Smith, in order to activate the wireless LAN card under a 64-bit version of
Linux, as acerhk[1] (the previous fine solution to the problem) relied on making
BIOS calls which are not allowed from a 64-bit OS.

Please read the INSTALL file for installation instructions.

acer_acpi is now being maintained by Carlos Corbacho <carlos@strangeworlds.co.uk>

[1] acerhk: http://www.cakey.de/acerhk/

Usage
*****

As root, modprobe acer_acpi

(To enable debugging, add debug=X, where X is between 0 and 2).

(This will also load the module 'wmi-acer', which is responsible for converting
ACPI-WMI calls to ACPI calls).

acer_acpi creates entries under /proc/acpi/acer (deprecated),
/sys/devices/platform/acer_acpi, /sys/class/backlight (2.6.20 and above only)
and /sys/class/led, which enables you to control some of the following (varies
between models):

* the wireless LAN card
* inbuilt Bluetooth adapter
* mail LED of your laptop
* brightness of the LCD panel

The entries for each device will only be created if the device exists on your
system.

The syntax for each is the same:

PLEASE NOTE: "enabled: X" support has now been removed

To read the status of a device (0=off, 1=on):
cat /proc/acpi/acer/{device}
or
cat /sys/devices/platform/acer_acpi/{device}

To enable a feature:
echo 1 > /proc/acpi/acer/{feature}
or
echo 1 > /sys/devices/platform/acer_acpi/{device}

To disable a feature:
echo 0 > /proc/acpi/acer/{feature}
echo 0 > /sys/devices/platform/acer_acpi/{device}

(brightness and mail LED are not visible in /sys/devices/platform - these are
properly registered with the relevant kernel subsystem).

To enable a feature on loading the module:
modprobe acer_acpi {feature}=1

To disable a feature on loading the module (default):
modprobe acer_acpi {feature}=0

Where {feature} is either wireless, bluetooth, threeg (aka 3G), or mailled.
(You can have multiple {feature}'s specified on the modprobe line).

For laptops with a "brightness" option, you can write a value between 0 to 9 or
15 (or 0x0 to 0x9 or 0xF - the maximum brightness varies between laptops).

Notes
*****

The behaviour of the wireless LED will depend on your hardware and driver
combination.

e.g. With the BCM4318 on the 5020 series:

ndiswrapper: Light blinks on when transmitting
bcm43xx/b43: Solid light, blinks off when transmitting

Credits
*******

Olaf Tauber, who did the real hard work when he developed acerhk
http://www.informatik.hu-berlin.de/~tauber/acerhk
All the authors of laptop acpi modules in the kernel, whose work
was an inspiration and a source of good code, particularly John Belmonte,
author of the toshiba_acpi driver.
Mathieu Segaud, who solved the problem with having to modprobe the driver
twice in version 0.2 and below.
Jim Ramsay, who added support for the new WMID interface

Contact
*******

If you have problems with the driver, please file a bug on the website and
include the following information:
1. name of your laptop (e.g. Acer Aspire 5021)
2. contents of your ACPI DSDT (cat /proc/acpi/dsdt).  I can disassemble
   it if necessary, so I'll accept the binaries.

Please check http://code.google.com/p/aceracpi/wiki/SupportedHardware first to
see if your hardware is supported or unsupported.

The latest version of acer_acpi can be found here:

http://code.google.com/p/aceracpi


NEWS
2008-03-11 Version 0.11.2
	Fix compile issue on older kernels
	Add EC quirks for 3610 & 5610
	Clean up documentation on 3G
	 Remove the warning on not detecting the mail LED

2008-02-08 Version 0.11.1
	Fix backlight on AMW0

2008-02-08 Version 0.11
	Convert AMW0 V2 devices to use WMID methods
	Enable Aspire 5020 quirks for all AMW0 (V1) laptops
	Add different EC quirk for Medion MD 98000 (AMW0)
	Add device/ feature autodetection
	Remove all the EC quirks that are no longer needed after the AMW0 V2 and
	autodetection changes.
	Simplify some of the internal code handling for WMID and AMW0 devices
	Fix build on OpenSuSE 10.3 with patched 2.6.22 kernel
	Fix mail LED return value on WMID
	Re-add INSTALL file

2007-12-16 Version 0.10
	Add ACPI-WMI mapper driver (wmi-acer)
	Convert acer_acpi to a platform driver, and call the WMI mapper driver
	Add EC support for Acer Aspire 5650, 3100, 5100 and 9420
	Fix EC support for Acer Aspire 5630 and 9300
	Fix build on 2.6.24 kernels
	Add DMI based autoloading for Acer laptops in 2.6.23 and newer kernels

2007-09-29 Version 0.9.1
	Add EC support for Acer TravelMate 5620
	Fix WMID
	Fix more read/ write errors

2007-09-27 Version 0.9.0
	Fix compiling on older (pre 2.6.20) kernels
	Add Acer Aspire 5560, TravelMate 2420 and TravelMate 5720 EC quirks
	Cleanup of interface code

2007-09-03 Version 0.8.2
	Really fix backlight code on older kernels
	Fix write errors with /proc/acpi/acer/brightness
	Fix problem with driver registration on WMID

2007-09-02 Version 0.8.1
	Fix backlight code on older (pre 2.6.21) kernels

2007-09-01 Version 0.8
	Register backlight with backlight driver
	Restored 3G support
	Added new sysfs platform driver to /sys/devices/platform/acer_acpi
	Added EC support for Aspire 3020, 3040, 5040, 5650 and 9300
	Code clean up - switched C99 types to kernel types, removed typedefs

2007-08-02 Version 0.7
	Dropped 'enabled: X' support
	Added limited direct EC support for Aspire 5020 and TravelMate 2490
	Fixed broken AMW0 values on hibernate
        Added new sysfs interface to /sys/devices/platform/acer_acpi

2007-07-17 Version 0.6.3
	Fix Bash 3.2 bug in Makefile (really fixes *buntu build bug)

2007-07-15 Version 0.6.2
	Fix packaging bug from 0.6.1 (directory name accidentally
	dropped acer_acpi-)

2007-07-13 Version 0.6.1
	Remove 2.4 kernel check, as it breaks *buntu systems

2007-07-06 Version 0.6
	Cleaned up code
	Added WMID support
	Makefile no longer gives harmless errors on older 2.6 kernels

2007-05-05 Version 0.5
	New maintainer added: Carlos Corbacho <cathectic@gmail.com>
	cat /proc/acpi/acer/{feature} now works - see ChangeLog for more

2007-03-31 Version 0.4
	Compiles under later 2.6 series kernels.
	New home at code.google.com

2005-08-30 Version 0.3
	Fixed a bug in the call to acpi_bus_register_driver, so the module
	no longer needs loading twice.

2005-08-28 Version 0.2
	First public release, this version actually does some useful things.

2005-08-27 Version 0.1
	Initial release that enables wireless on modprobe.  And that's it.


wmi-acer.h
/*
 *  wmi.h
 *
 *  Copyright (C) 2007 Carlos Corbacho <cathectic@gmail.com>
 *
 * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or (at
 *  your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful, but
 *  WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 *
 * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 */

/* Workaround needed for older kernels */
#ifndef bool
#define bool int
#define true    1
#define false   0
#endif

typedef void (*wmi_notify_handler) (u32 value, void *context);

extern acpi_status wmi_acer_evaluate_method(const char *guid, u8 instance,
					u32 method_id,
					const struct acpi_buffer *in,
					struct acpi_buffer *out);
extern acpi_status wmi_acer_query_block(const char *guid, u8 instance,
					struct acpi_buffer *out);
extern acpi_status wmi_acer_set_block(const char *guid, u8 instance,
					const struct acpi_buffer *in);
extern acpi_status wmi_acer_install_notify_handler(wmi_notify_handler handler,
					void *data);
extern acpi_status wmi_acer_remove_notify_handler(void);
extern acpi_status wmi_acer_get_event_data(u32 event, struct acpi_buffer *out);
extern bool wmi_acer_has_guid(const char *guid);


wmi-acer.c
/*
 *  ACPI-WMI mapping driver (acer_acpi port)
 *
 *  Copyright (C) 2007 Carlos Corbacho <carlos@strangeworlds.co.uk>
 *
 *  GUID parsing code from ldm.c is:
 *   Copyright (C) 2001,2002 Richard Russon <ldm@flatcap.org>
 *   Copyright (c) 2001-2007 Anton Altaparmakov
 *   Copyright (C) 2001,2002 Jakob Kemi <jakob.kemi@telia.com>
 *
 * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or (at
 *  your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful, but
 *  WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 *
 * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 */

#include <linux/kernel.h>
#include <linux/module.h>
#include <linux/version.h>
#include <linux/init.h>
#include <linux/types.h>
#include <linux/list.h>

#include <acpi/acpi_bus.h>
#include <acpi/acpi_drivers.h>

#include "wmi-acer.h"

#define ACPI_WMI_CLASS			"wmi-acer"

#undef PREFIX
#define PREFIX "ACPI: WMI-Acer: "
#define WMI_ERR KERN_ERR PREFIX
#define WMI_NOTICE KERN_NOTICE PREFIX
#define WMI_INFO KERN_INFO PREFIX

#define DEBUG(level, message...) { \
	if (debug >= level) \
		printk(KERN_DEBUG PREFIX message);\
}

MODULE_AUTHOR("Carlos Corbacho");
MODULE_DESCRIPTION("ACPI-WMI Mapping Driver - acer_acpi port");
MODULE_LICENSE("GPL");

static DEFINE_MUTEX(wmi_data_lock);

struct guid_block
{
	char guid[16];
	union
	{
		char object_id[2];
		struct
		{
			unsigned char notify_id;
			unsigned char reserved;
		};
	};
	u8 instance_count;
	u8 flags;
};

struct wmi_block
{
	struct list_head list;
	struct guid_block gblock;
	acpi_handle handle;
};

static struct wmi_block wmi_blocks;

static wmi_notify_handler wmi_external_handler;
static void *wmi_external_data;

static int debug;

module_param(debug, int, 0664);
MODULE_PARM_DESC(debug, "Debugging verbosity level (0=least 2=most)");

/*
 * If the GUID data block is marked as expensive, we must enable and
 * explicitily disable data collection.
 */
#define ACPI_WMI_EXPENSIVE   0x1
#define ACPI_WMI_METHOD      0x2	/* GUID is a method */
#define ACPI_WMI_STRING      0x4	/* GUID takes & returns a string */
#define ACPI_WMI_EVENT       0x8	/* GUID is an event */

static int acpi_wmi_remove(struct acpi_device *device, int type);
static int acpi_wmi_add(struct acpi_device *device);

#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,23)) || \
(defined(CONFIG_SUSE_KERNEL) && LINUX_VERSION_CODE == KERNEL_VERSION(2,6,22))
const static struct acpi_device_id wmi_device_ids[] = {
	{"PNP0C14", 0},
	{"pnp0c14", 0},
	{"", 0},
};
MODULE_DEVICE_TABLE(acpi, wmi_device_ids);
#endif

static struct acpi_driver acpi_wmi_driver = {
	.name = "wmi-acer",
	.class = ACPI_WMI_CLASS,
#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,23)) || \
(defined(CONFIG_SUSE_KERNEL) && LINUX_VERSION_CODE == KERNEL_VERSION(2,6,22))
	.ids = wmi_device_ids,
#else
	.ids = "PNP0C14,pnp0c14",
#endif
	.ops = {
		.add = acpi_wmi_add,
		.remove = acpi_wmi_remove,
		},
};

/*
 * GUID parsing functions
 */

/**
 * wmi_parse_hexbyte - Convert a ASCII hex number to a byte
 * @src:  Pointer to at least 2 characters to convert.
 *
 * Convert a two character ASCII hex string to a number.
 *
 * Return:  0-255  Success, the byte was parsed correctly
 *          -1     Error, an invalid character was supplied
 */
static int wmi_parse_hexbyte(const u8 *src)
{
	unsigned int x; /* For correct wrapping */
	int h;

	/* high part */
	x = src[0];
	if (x - '0' <= '9' - '0') {
		h = x - '0';
	} else if (x - 'a' <= 'f' - 'a') {
		h = x - 'a' + 10;
	} else if (x - 'A' <= 'F' - 'A') {
		h = x - 'A' + 10;
	} else {
		return -1;
	}
	h <<= 4;

	/* low part */
	x = src[1];
	if (x - '0' <= '9' - '0')
		return h | (x - '0');
	if (x - 'a' <= 'f' - 'a')
		return h | (x - 'a' + 10);
	if (x - 'A' <= 'F' - 'A')
		return h | (x - 'A' + 10);
	return -1;
}

/**
 * wmi_swap_bytes - Rearrange GUID bytes to match GUID binary
 * @src:   Memory block holding binary GUID (16 bytes)
 * @dest:  Memory block to hold byte swapped binary GUID (16 bytes)
 *
 * Byte swap a binary GUID to match it's real GUID value
 */
static void wmi_swap_bytes(u8 *src, u8 *dest)
{
	int i;

	for (i = 0; i <= 3; i++)
		memcpy(dest + i, src + (3 - i), 1);

	for (i = 0; i <= 1; i++)
		memcpy(dest + 4 + i, src + (5 - i), 1);

	for (i = 0; i <= 1; i++)
		memcpy(dest + 6 + i, src + (7 - i), 1);

	memcpy(dest + 8, src + 8, 8);
}

/**
 * wmi_parse_guid - Convert GUID from ASCII to binary
 * @src:   36 char string of the form fa50ff2b-f2e8-45de-83fa-65417f2f49ba
 * @dest:  Memory block to hold binary GUID (16 bytes)
 *
 * N.B. The GUID need not be NULL terminated.
 *
 * Return:  'true'   @dest contains binary GUID
 *          'false'  @dest contents are undefined
 */
static bool wmi_parse_guid(const u8 *src, u8 *dest)
{
	static const int size[] = { 4, 2, 2, 2, 6 };
	int i, j, v;

	if (src[8]  != '-' || src[13] != '-' ||
		src[18] != '-' || src[23] != '-')
		return false;

	for (j = 0; j < 5; j++, src++) {
		for (i = 0; i < size[j]; i++, src += 2, *dest++ = v) {
			v = wmi_parse_hexbyte(src);
			if (v < 0)
				return false;
		}
	}

	return true;
}

static bool find_guid(const char *guid_string, struct wmi_block **out)
{
	char tmp[16], guid_input[16];
	struct wmi_block *wblock;
	struct guid_block *block;
	struct list_head *p;

	DEBUG(2, "find_guid called\n");
	DEBUG(2, "Passing GUID to parser\n");
	wmi_parse_guid(guid_string, tmp);
	wmi_swap_bytes(tmp, guid_input);

	list_for_each(p, &wmi_blocks.list) {
		wblock = list_entry(p, struct wmi_block, list);
		block = &wblock->gblock;

		if (memcmp(block->guid, guid_input, 16) == 0) {
			DEBUG(2, "GUID found - returning block\n");
			if (out)
				*out = wblock;
			return 1;
		}
	}
	return 0;
}

/*
 * Exported WMI functions
 */
/**
 * wmi_acer_evaluate_method - Evaluate a WMI method
 * @guid_string: 36 char string of the form fa50ff2b-f2e8-45de-83fa-65417f2f49ba
 * @instance: Instance index
 * @method_id: Method ID to call
 * &in: Buffer containing input for the method call
 * &out: Empty buffer to return the method results
 *
 * Convert a WMI method call to an ACPI one, and return the results
 */
acpi_status wmi_acer_evaluate_method(const char *guid_string, u8 instance,
u32 method_id, const struct acpi_buffer *in, struct acpi_buffer *out)
{
	struct guid_block *block = NULL;
	struct wmi_block *wblock = NULL;
	acpi_handle handle;
	acpi_status status;
	struct acpi_object_list input;
	union acpi_object params[3];
	char method[4] = "WM";

	if (!find_guid(guid_string, &wblock))
		return AE_BAD_ADDRESS;

	block = &wblock->gblock;
	handle = wblock->handle;

	if (!block->flags & ACPI_WMI_METHOD) {
		DEBUG(2, "GUID is not a method\n");
		return AE_BAD_DATA;
	}

	if (block->instance_count < instance)
		return AE_BAD_PARAMETER;

	input.count = 2;
	input.pointer = params;
	params[0].type = ACPI_TYPE_INTEGER;
	params[0].integer.value = instance;
	params[1].type = ACPI_TYPE_INTEGER;
	params[1].integer.value = method_id;

	if (in != NULL) {
		input.count = 3;

		if (block->flags & ACPI_WMI_STRING) {
			params[2].type = ACPI_TYPE_STRING;
		} else {
			params[2].type = ACPI_TYPE_BUFFER;
		}
		params[2].buffer.length = in->length;
		params[2].buffer.pointer = in->pointer;
	}

	strncat(method, block->object_id, 2);
	DEBUG(2, "Object to call is %s\n", method);

	status = acpi_evaluate_object(handle, method, &input, out);

	return status;
}
EXPORT_SYMBOL_GPL(wmi_acer_evaluate_method);

/**
 * wmi_query_block - Return contents of a WMI block
 * @guid_string: 36 char string of the form fa50ff2b-f2e8-45de-83fa-65417f2f49ba
 * @instance: Instance index
 * &out: Empty buffer to return the contents of the data block to
 *
 * Return the contents of a data block to a buffer
 */
acpi_status wmi_acer_query_block(const char *guid_string, u8 instance,
struct acpi_buffer *out)
{
	struct guid_block *block = NULL;
	struct wmi_block *wblock = NULL;
	acpi_handle handle;
	acpi_status status, wc_status = AE_ERROR;
	struct acpi_object_list input, wc_input;
	union acpi_object wc_params[1], wq_params[1];
	char method[4];
	char wc_method[4] = "WC";

	if (guid_string == NULL || out == NULL)
		return AE_BAD_PARAMETER;

	if (!find_guid(guid_string, &wblock))
		return AE_BAD_ADDRESS;

	block = &wblock->gblock;
	handle = wblock->handle;

	if (block->instance_count < instance)
		return AE_BAD_PARAMETER;

	/* Check GUID is a data block */
	if (block->flags & (ACPI_WMI_EVENT | ACPI_WMI_METHOD))
		return AE_BAD_ADDRESS;

	input.count = 1;
	input.pointer = wq_params;
	wq_params[0].type = ACPI_TYPE_INTEGER;
	wq_params[0].integer.value = instance;

	/*
	 * If ACPI_WMI_EXPENSIVE, call the relevant WCxx method first to
	 * enable collection
	 */
	if (block->flags & ACPI_WMI_EXPENSIVE) {
		wc_input.count = 1;
		wc_input.pointer = wc_params;
		wc_params[0].type = ACPI_TYPE_INTEGER;
		wc_params[0].integer.value = 1;

		strncat(wc_method, block->object_id, 2);

		/*
		 * Some GUIDs break the specification by declaring themselves
		 * expensive, but have no corresponding WCxx method. So we
		 * should not fail if this happens.
		 */
		wc_status = acpi_evaluate_object(handle, wc_method,
			&wc_input, NULL);
	}

	strcpy(method, "WQ");
	strncat(method, block->object_id, 2);
	DEBUG(2, "Object to call is %s\n", method);

	status = acpi_evaluate_object(handle, method, NULL, out);

	/*
	 * If ACPI_WMI_EXPENSIVE, call the relevant WCxx method, even if
	 * the WQxx method failed - we should disable collection anyway
	 */
	if ((block->flags & ACPI_WMI_EXPENSIVE) && ACPI_SUCCESS(wc_status)) {
		wc_params[0].integer.value = 0;
		status = acpi_evaluate_object(handle,
		wc_method, &wc_input, NULL);
	}

	return status;
}
EXPORT_SYMBOL_GPL(wmi_acer_query_block);

/**
 * wmi_acer_set_block - Write to a WMI block
 * @guid_string: 36 char string of the form fa50ff2b-f2e8-45de-83fa-65417f2f49ba
 * @instance: Instance index
 * &in: Buffer containing new values for the data block
 *
 * Write the contents of the input buffer to ACPI
 */
acpi_status wmi_acer_set_block(const char *guid_string, u8 instance,
const struct acpi_buffer *in)
{
	struct guid_block *block = NULL;
	struct wmi_block *wblock = NULL;
	acpi_handle handle;
	struct acpi_object_list input;
	union acpi_object params[2];
	char method[4] = "WS";

	if (guid_string == NULL || in == NULL)
		return AE_BAD_DATA;

	if (!find_guid(guid_string, &wblock))
		return AE_BAD_ADDRESS;

	block = &wblock->gblock;
	handle = wblock->handle;

	if (block->instance_count < instance)
		return AE_BAD_PARAMETER;

	/* Check GUID is a data block */
	if (block->flags & (ACPI_WMI_EVENT | ACPI_WMI_METHOD))
		return AE_BAD_ADDRESS;

	input.count = 2;
	input.pointer = params;
	params[0].type = ACPI_TYPE_INTEGER;
	params[0].integer.value = instance;

	if (block->flags & ACPI_WMI_STRING) {
		params[1].type = ACPI_TYPE_STRING;
	} else {
		params[1].type = ACPI_TYPE_BUFFER;
	}
	params[1].buffer.length = in->length;
	params[1].buffer.pointer = in->pointer;

	strncat(method, block->object_id, 2);
	DEBUG(2, "Object to call is %s\n", method);

	return acpi_evaluate_object(handle, method, &input, NULL);
}
EXPORT_SYMBOL_GPL(wmi_acer_set_block);

/**
 * wmi_install_notify_handler - Register handler for WMI events
 * @handler: Function to handle notifications
 * @data: Data to be returned to handler when event is fired
 *
 * Register a handler for events sent to the WMI-ACPI mapper device.
 */
acpi_status wmi_acer_install_notify_handler(wmi_notify_handler handler, void *data)
{
	if (!handler)
		return AE_BAD_PARAMETER;

	if (!wmi_external_handler)
		return AE_ALREADY_ACQUIRED;

	wmi_external_handler = handler;
	wmi_external_data = data;

	return AE_OK;
}
EXPORT_SYMBOL_GPL(wmi_acer_install_notify_handler);

/**
 * wmi_acer_uninstall_notify_handler - Unregister handler for WMI events
 *
 * Unregister handler for events sent to the WMI-ACPI mapper device.
 */
acpi_status wmi_acer_remove_notify_handler(void)
{
	DEBUG(1, "Notifier triggered\n");

	if (wmi_external_handler) {
		wmi_external_handler = NULL;
		wmi_external_data = NULL;
		return AE_OK;
	}
	return AE_ERROR;
}
EXPORT_SYMBOL_GPL(wmi_acer_remove_notify_handler);

/**
 * wmi_acer_get_event_data - Get WMI data associated with an event
 *
 * @event - Event to find
 * &out - Buffer to hold event data
 *
 * Returns extra data associated with an event in WMI.
 */
acpi_status wmi_acer_get_event_data(u32 event, struct acpi_buffer *out)
{
	struct acpi_object_list input;
	union acpi_object params[1];
	struct guid_block *gblock;
	struct wmi_block *wblock;
	struct list_head *p;

	input.count = 1;
	input.pointer = params;
	params[0].type = ACPI_TYPE_INTEGER;
	params[0].integer.value = event;

	list_for_each(p, &wmi_blocks.list) {
		wblock = list_entry(p, struct wmi_block, list);
		gblock = &wblock->gblock;

		if ((gblock->flags & ACPI_WMI_EVENT) &&
			(gblock->notify_id == event))
			return acpi_evaluate_object(wblock->handle, "_WED",
				&input, out);
	}

	return AE_NOT_FOUND;
}
EXPORT_SYMBOL(wmi_acer_get_event_data);

/**
 * wmi_acer_has_guid - Check if a GUID is available
 * @guid_string: 36 char string of the form fa50ff2b-f2e8-45de-83fa-65417f2f49ba
 *
 * Check if a given GUID is defined by _WDG
 */
bool wmi_acer_has_guid(const char *guid_string)
{
	return find_guid(guid_string, NULL);
}
EXPORT_SYMBOL_GPL(wmi_acer_has_guid);

/**
 * parse_wdg - Parse the _WDG method for the GUID data blocks
 */
static __init acpi_status parse_wdg(acpi_handle handle)
{
	struct acpi_buffer out = {ACPI_ALLOCATE_BUFFER, NULL};
	union acpi_object *obj;
	struct guid_block *gblock;
	struct wmi_block *wblock;
	acpi_status status;
	u32 i, total;

	status = acpi_evaluate_object(handle, "_WDG", NULL, &out);

	if (ACPI_FAILURE(status))
		return status;

	obj = (union acpi_object *) out.pointer;

	if (obj->type != ACPI_TYPE_BUFFER)
		return AE_ERROR;

	total = obj->buffer.length / sizeof(struct guid_block);

	gblock = kzalloc(obj->buffer.length, GFP_KERNEL);
	if (!gblock)
		return AE_NO_MEMORY;

	memcpy(gblock, obj->buffer.pointer, obj->buffer.length);

	for (i = 0; i < total; i++) {
		wblock = kzalloc(sizeof(struct wmi_block), GFP_KERNEL);
		if (!wblock)
			return AE_NO_MEMORY;

		wblock->gblock = gblock[i];
		wblock->handle = handle;
		list_add_tail(&wblock->list, &wmi_blocks.list);
	}

	kfree(out.pointer);
	kfree(gblock);

	return status;
}

/*
 * WMI can have EmbeddedControl access regions. In which case, we just want to
 * hand these off to the EC driver.
 */
static acpi_status
acpi_wmi_ec_space_handler(u32 function, acpi_physical_address address,
		      u32 bits, acpi_integer *value,
		      void *handler_context, void *region_context)
{
	int result = 0, i = 0;
	u8 temp = 0;

	if ((address > 0xFF) || !value)
		return AE_BAD_PARAMETER;

	if (function != ACPI_READ && function != ACPI_WRITE)
		return AE_BAD_PARAMETER;

	if (bits != 8)
		return AE_BAD_PARAMETER;

	if (function == ACPI_READ) {
		result = ec_read(address, &temp);
		(*value) |= ((acpi_integer)temp) << i;
	} else {
		temp = 0xff & ((*value) >> i);
		result = ec_write(address, temp);
	}

	switch (result) {
	case -EINVAL:
		return AE_BAD_PARAMETER;
		break;
	case -ENODEV:
		return AE_NOT_FOUND;
		break;
	case -ETIME:
		return AE_TIME;
		break;
	default:
		return AE_OK;
	}
}

static void acpi_wmi_notify(acpi_handle handle, u32 event, void *data)
{
	struct guid_block *block;
	struct wmi_block *wblock;
	struct list_head *p;

	list_for_each(p, &wmi_blocks.list) {
		wblock = list_entry(p, struct wmi_block, list);
		block = &wblock->gblock;

		if ((block->flags & ACPI_WMI_EVENT) &&
		block->notify_id == event) {
			if (wmi_external_handler)
				wmi_external_handler(event, wmi_external_data);
			break;
		}
	}
}

static int __init acpi_wmi_add(struct acpi_device *device)
{
	acpi_status status;
	int result = 0;

	status = acpi_install_notify_handler(device->handle, ACPI_DEVICE_NOTIFY,
		acpi_wmi_notify, device);

	if (ACPI_FAILURE(status)) {
		ACPI_DEBUG_PRINT((ACPI_DB_ERROR,
				  "Error installing notify handler\n"));
		return -ENODEV;
	}

	status = acpi_install_address_space_handler(device->handle,
						    ACPI_ADR_SPACE_EC,
						    &acpi_wmi_ec_space_handler,
						    NULL, NULL);
	if (ACPI_FAILURE(status)) {
		ACPI_DEBUG_PRINT((ACPI_DB_ERROR,
				  "Error installing EC region handler\n"));
		return -ENODEV;
	}

	status = parse_wdg(device->handle);
	if (ACPI_FAILURE(status))
		return -ENODEV;

	return result;
}

static int acpi_wmi_remove(struct acpi_device *device, int type)
{
	acpi_remove_address_space_handler(device->handle, ACPI_ADR_SPACE_EC,
						    &acpi_wmi_ec_space_handler);

	acpi_remove_notify_handler(device->handle, ACPI_DEVICE_NOTIFY,
		acpi_wmi_notify);

	return 0;
}

static int __init acpi_wmi_init(void)
{
	acpi_status result;

	if (acpi_disabled)
		return -ENODEV;

	INIT_LIST_HEAD(&wmi_blocks.list);

	result = acpi_bus_register_driver(&acpi_wmi_driver);

	if (result < 0) {
		printk(KERN_INFO PREFIX "Interface device not found\n");
		return result;
	}

	printk(KERN_INFO PREFIX "Mapper loaded\n");

	return result;
}

static void __exit acpi_wmi_exit(void)
{
	struct list_head *p, *tmp;
	struct wmi_block *wblock;

	acpi_bus_unregister_driver(&acpi_wmi_driver);

	list_for_each_safe(p, tmp, &wmi_blocks.list) {
		wblock = list_entry(p, struct wmi_block, list);

		list_del(p);
		kfree(wblock);
	}

	printk(KERN_INFO PREFIX "Mapper unloaded\n");
}

module_init(acpi_wmi_init);
module_exit(acpi_wmi_exit);
